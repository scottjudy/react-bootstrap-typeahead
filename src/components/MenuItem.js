// @flow

import cx from 'classnames';
import React, { type Node } from 'react';

import { withItem } from '../behaviors/item';

import type { EventHandler } from '../types';

type MenuItemProps = {
  active?: boolean,
  children?: Node,
  className?: string,
  disabled?: boolean,
  href?: string,
  onClick?: EventHandler<HTMLElement>,
};

const BaseMenuItem = React.forwardRef<MenuItemProps, ?HTMLElement>((
  { active, children, className, disabled, onClick, ...props },
  ref
) => {
  return (
    <a
      {...props}
      className={cx('dropdown-item', { active, disabled }, className)}
      href={props.href || '#'}
      onClick={(e: SyntheticEvent<HTMLElement>) => {
        e.preventDefault();
        !disabled && onClick && onClick(e);
      }}
      ref={ref}>
      {children}
    </a>
  );
});

export { BaseMenuItem };
export default withItem(BaseMenuItem);
