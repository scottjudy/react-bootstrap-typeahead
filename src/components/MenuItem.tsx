import cx from 'classnames';
import React, { forwardRef, HTMLProps, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../behaviors/item';

export interface MenuItemProps extends HTMLProps<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, MenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export default function MenuItem(props: UseItemProps<HTMLAnchorElement>) {
  return <BaseMenuItem {...useItem(props)} />;
}
