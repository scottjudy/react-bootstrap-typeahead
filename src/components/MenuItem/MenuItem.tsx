import cx from 'classnames';
import React, { forwardRef, HTMLAttributes, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../../behaviors/item';

export interface BaseMenuItemProps extends HTMLAttributes<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
  href?: string;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, BaseMenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export type MenuItemProps = UseItemProps<HTMLAnchorElement>;

export default function MenuItem(props: MenuItemProps) {
  return <BaseMenuItem {...useItem(props)} />;
}
