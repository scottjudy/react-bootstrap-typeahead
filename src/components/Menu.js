// @flow

import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { Children, type Node } from 'react';

import { BaseMenuItem } from './MenuItem';

import { preventInputBlur } from '../utils';
import { checkPropType, isRequiredForA11y } from '../propTypes';
import type { Id, MenuProps } from '../types';

const MenuDivider = (props: any) => (
  <div className="dropdown-divider" role="separator" />
);

const MenuHeader = (props: any) => (
  // eslint-disable-next-line jsx-a11y/role-has-required-aria-props
  <div {...props} className="dropdown-header" role="heading" />
);

const propTypes = {
  'aria-label': PropTypes.string,
  /**
   * Message to display in the menu if there are no valid results.
   */
  emptyLabel: PropTypes.node,
  /**
   * Needed for accessibility.
   */
  id: checkPropType(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    isRequiredForA11y,
  ),
  /**
   * Maximum height of the dropdown menu.
   */
  maxHeight: PropTypes.string,
};

const defaultProps = {
  'aria-label': 'menu-options',
  emptyLabel: 'No matches found.',
  maxHeight: '300px',
};

export type MenuComponentProps = MenuProps & {
  'aria-label': string,
  children?: Node,
  className?: string,
  emptyLabel: Node,
  id: Id,
  maxHeight: string,
  text: string,
};

/**
 * Menu component that handles empty state when passed a set of results.
 */
class Menu extends React.Component<MenuComponentProps> {
  static propTypes = propTypes;
  static defaultProps = defaultProps;
  static Divider = MenuDivider;
  static Header = MenuHeader;

  componentDidUpdate(prevProps: MenuComponentProps) {
    const { inputHeight, scheduleUpdate } = this.props;

    // Update the menu position if the height of the input changes.
    if (inputHeight !== prevProps.inputHeight) {
      scheduleUpdate();
    }
  }

  render() {
    const {
      children,
      className,
      emptyLabel,
      id,
      innerRef,
      maxHeight,
      style,
      text,
    } = this.props;

    const contents = Children.count(children) === 0 ?
      <BaseMenuItem disabled role="option">
        {emptyLabel}
      </BaseMenuItem> :
      children;

    return (
      /* eslint-disable jsx-a11y/interactive-supports-focus */
      <div
        aria-label={this.props['aria-label']}
        className={cx('rbt-menu', 'dropdown-menu', 'show', className)}
        id={id}
        key={
          // Force a re-render if the text changes to ensure that menu
          // positioning updates correctly.
          text
        }
        onMouseDown={
          // Prevent input from blurring when clicking on the menu scrollbar.
          preventInputBlur
        }
        ref={innerRef}
        role="listbox"
        style={{
          ...style,
          display: 'block',
          maxHeight,
          overflow: 'auto',
        }}>
        {contents}
      </div>
      /* eslint-enable jsx-a11y/interactive-supports-focus */
    );
  }
}

export default Menu;
