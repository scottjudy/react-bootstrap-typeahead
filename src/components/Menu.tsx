import cx from 'classnames';
import PropTypes from 'prop-types';
import React, {
  Children,
  HTMLProps,
  ReactNode,
  Ref,
  useEffect,
  useRef,
} from 'react';

import { BaseMenuItem } from './MenuItem';

import { preventInputBlur } from '../utils';
import { checkPropType, isRequiredForA11y } from '../propTypes';

const MenuDivider = () => <div className="dropdown-divider" role="separator" />;

const MenuHeader = (props: HTMLProps<HTMLDivElement>) => (
  // eslint-disable-next-line jsx-a11y/role-has-required-aria-props
  <div {...props} className="dropdown-header" role="heading" />
);

const propTypes = {
  'aria-label': PropTypes.string,
  /**
   * Message to display in the menu if there are no valid results.
   */
  emptyLabel: PropTypes.node,
  /**
   * Needed for accessibility.
   */
  id: checkPropType(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    isRequiredForA11y
  ),
  /**
   * Maximum height of the dropdown menu.
   */
  maxHeight: PropTypes.string,
};

const defaultProps = {
  'aria-label': 'menu-options',
};

export interface MenuProps extends HTMLProps<HTMLDivElement> {
  emptyLabel?: ReactNode;
  innerRef: Ref<HTMLDivElement>;
  inputHeight: number;
  scheduleUpdate: () => void;
  maxHeight?: string;
  text: string;
}

/**
 * Menu component that handles empty state when passed a set of results.
 */
const Menu = ({
  children,
  className,
  emptyLabel = 'No matches found.',
  innerRef,
  inputHeight,
  maxHeight = '300px',
  scheduleUpdate,
  style,
  text,
  ...props
}: MenuProps) => {
  const isInitialMount = useRef(true);

  useEffect(() => {
    if (isInitialMount.current) {
      isInitialMount.current = false;
      return;
    }

    scheduleUpdate && scheduleUpdate();
  }, [inputHeight, scheduleUpdate]);

  const contents =
    Children.count(children) === 0 ? (
      <BaseMenuItem disabled role="option">
        {emptyLabel}
      </BaseMenuItem>
    ) : (
      children
    );

  return (
    /* eslint-disable jsx-a11y/interactive-supports-focus */
    <div
      {...props}
      className={cx('rbt-menu', 'dropdown-menu', 'show', className)}
      key={
        // Force a re-render if the text changes to ensure that menu
        // positioning updates correctly.
        text
      }
      onMouseDown={
        // Prevent input from blurring when clicking on the menu scrollbar.
        preventInputBlur
      }
      ref={innerRef}
      role="listbox"
      style={{
        ...style,
        display: 'block',
        maxHeight,
        overflow: 'auto',
      }}>
      {contents}
    </div>
    /* eslint-enable jsx-a11y/interactive-supports-focus */
  );
};

Menu.propTypes = propTypes;
Menu.defaultProps = defaultProps;
Menu.Divider = MenuDivider;
Menu.Header = MenuHeader;

export default Menu;
