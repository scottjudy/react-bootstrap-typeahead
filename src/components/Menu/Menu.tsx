import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { Children, HTMLProps, ReactNode, Ref } from 'react';

import { BaseMenuItem } from '../MenuItem';

import { preventInputBlur } from '../../utils';
import { checkPropType, isRequiredForA11y } from '../../propTypes';

const MenuDivider = () => <div className="dropdown-divider" role="separator" />;

const MenuHeader = (props: HTMLProps<HTMLDivElement>) => (
  // eslint-disable-next-line jsx-a11y/role-has-required-aria-props
  <div {...props} className="dropdown-header" role="heading" />
);

const propTypes = {
  'aria-label': PropTypes.string,
  /**
   * Message to display in the menu if there are no valid results.
   */
  emptyLabel: PropTypes.node,
  /**
   * Needed for accessibility.
   */
  id: checkPropType(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    isRequiredForA11y
  ),
  /**
   * Maximum height of the dropdown menu.
   */
  maxHeight: PropTypes.string,
};

const defaultProps = {
  'aria-label': 'menu-options',
};

export interface MenuProps extends HTMLProps<HTMLDivElement> {
  emptyLabel?: ReactNode;
  innerRef?: Ref<HTMLDivElement>;
  maxHeight?: string;
  text?: string;
}

/**
 * Menu component that handles empty state when passed a set of results.
 */
const Menu = ({
  emptyLabel = 'No matches found.',
  innerRef,
  maxHeight = '300px',
  style,
  text = '',
  ...props
}: MenuProps) => {
  const children =
    Children.count(props.children) === 0 ? (
      <BaseMenuItem disabled role="option">
        {emptyLabel}
      </BaseMenuItem>
    ) : (
      props.children
    );

  return (
    /* eslint-disable jsx-a11y/interactive-supports-focus */
    <div
      {...props}
      className={cx('rbt-menu', 'dropdown-menu', 'show', props.className)}
      key={
        // Force a re-render if the text changes to ensure that menu
        // positioning updates correctly.
        text
      }
      onMouseDown={
        // Prevent input from blurring when clicking on the menu scrollbar.
        preventInputBlur
      }
      ref={innerRef}
      role="listbox"
      style={{
        ...style,
        display: 'block',
        maxHeight,
        overflow: 'auto',
      }}>
      {children}
    </div>
    /* eslint-enable jsx-a11y/interactive-supports-focus */
  );
};

Menu.propTypes = propTypes;
Menu.defaultProps = defaultProps;
Menu.Divider = MenuDivider;
Menu.Header = MenuHeader;

export default Menu;
