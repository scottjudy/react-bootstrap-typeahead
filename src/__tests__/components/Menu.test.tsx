import React from 'react';

import Menu from '../../components/Menu';
import MenuItem from '../../components/MenuItem';

import {
  fireEvent,
  getItems,
  getMenu,
  noop,
  prepareSnapshot,
  render,
  screen,
} from '../helpers';

const options = [{ label: 'Item 1' }, { label: 'Item 2' }, { label: 'Item 3' }];

const TestComponent = (props) => (
  <Menu id="menu-id" paginate={false} scheduleUpdate={noop} {...props}>
    {options.map((o, idx) => (
      <MenuItem key={o.label} option={o} position={idx}>
        {o.label}
      </MenuItem>
    ))}
  </Menu>
);

describe('<Menu>', () => {
  it('renders a snapshot', () => {
    expect(prepareSnapshot(<TestComponent />)).toMatchSnapshot();
  });

  it('renders a basic menu with menu items', () => {
    render(<TestComponent />);

    expect(getMenu(screen)).toHaveClass('rbt-menu dropdown-menu');
    expect(getItems(screen)).toHaveLength(3);
  });

  it('sets the maxHeight and other styles', () => {
    render(
      <TestComponent maxHeight="100px" style={{ backgroundColor: 'red' }} />
    );

    const menu = getMenu(screen);
    expect(menu).toHaveStyle('background-color: red');
    expect(menu).toHaveStyle('max-height: 100px');
  });

  it('renders an empty label when there are no children', () => {
    const emptyLabel = 'No matches.';
    render(<TestComponent emptyLabel={emptyLabel} />);

    const items = getItems(screen);
    expect(items).toHaveLength(1);
    expect(items[0]).toHaveClass('disabled');
    expect(items[0]).toHaveTextContent(emptyLabel);
  });

  it('adds an aria-label attribute to the menu', () => {
    render(<TestComponent aria-label="custom-label" />);
    expect(getMenu(screen)).toHaveAttribute('aria-label', 'custom-label');
  });

  it('updates the menu position if the input height changes', () => {
    const scheduleUpdate = jest.fn();
    const { rerender } = render(
      <TestComponent scheduleUpdate={scheduleUpdate} />
    );

    expect(scheduleUpdate).toHaveBeenCalledTimes(0);

    rerender(<TestComponent inputHeight={1} scheduleUpdate={scheduleUpdate} />);
    expect(scheduleUpdate).toHaveBeenCalledTimes(1);
  });

  it('prevents the input from blurring on mousedown', () => {
    render(<TestComponent />);

    const menu = getMenu(screen);

    // `false` means e.preventDefault was called.
    expect(fireEvent.mouseDown(menu)).toBe(false);
  });
});

it('<Menu.Divider>', () => {
  render(<Menu.Divider />);

  const divider = screen.getByRole('separator');
  expect(divider.tagName).toBe('DIV');
  expect(divider).toHaveClass('dropdown-divider');
});

it('<Menu.Header>', () => {
  const children = 'This is a menu header';
  render(<Menu.Header>{children}</Menu.Header>);

  const header = screen.getByRole('heading');
  expect(header.tagName).toBe('DIV');
  expect(header).toHaveClass('dropdown-header');
  expect(header).toHaveTextContent(children);
});
