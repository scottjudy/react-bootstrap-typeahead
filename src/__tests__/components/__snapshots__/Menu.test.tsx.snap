// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu> renders a snapshot 1`] = `
<div
  aria-label="menu-options"
  className="rbt-menu dropdown-menu show"
  id="menu-id"
  onMouseDown={[Function]}
  paginate={false}
  role="listbox"
  style={
    Object {
      "display": "block",
      "maxHeight": "300px",
      "overflow": "auto",
    }
  }
>
  <a
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-0"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "label": "Item 1",
      }
    }
    position={0}
    role="option"
  >
    Item 1
  </a>
  <a
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-1"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "label": "Item 2",
      }
    }
    position={1}
    role="option"
  >
    Item 2
  </a>
  <a
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-2"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "label": "Item 3",
      }
    }
    position={2}
    role="option"
  >
    Item 3
  </a>
</div>
`;
